src/app/components/autocomplete/autocomplete.component.ts

```diff
@@ -0,0 +1,149 @@
+import {
+	AbstractControl,
+	ControlValueAccessor,
+	FormControl,
+	NG_VALIDATORS,
+	NG_VALUE_ACCESSOR,
+	ValidationErrors,
+	Validators,
+} from '@angular/forms';
+import { combineLatest, debounceTime, map, Observable, of, startWith, switchMap } from 'rxjs';
+import { ErrorHandlerService, NpmRegistryService } from 'src/app/services';
+
+@Component({
+	selector: 'app-autocomplete',
+	templateUrl: './autocomplete.component.html',
+	styleUrls: ['./autocomplete.component.scss'],
+	changeDetection: ChangeDetectionStrategy.OnPush,
+	providers: [
+		{
+			provide: NG_VALUE_ACCESSOR,
+			useExisting: forwardRef(() => AutocompleteComponent),
+			multi: true,
+		},
+		{
+			provide: NG_VALIDATORS,
+			useExisting: forwardRef(() => AutocompleteComponent),
+			multi: true,
+		},
+	],
+})
+export class AutocompleteComponent implements ControlValueAccessor, Validators {
+	/*
+    keep track of package names that has been already loaded and show them
+    as suggestion for the user
+  */
+	@Input() autocompletePackageNames: string[] | null = [];
+
+	/*
+    Pacakge names that are already loaded to prevent selecting same library
+    multiple times
+  */
+	private alreadyLoadedPackageName: string[] = [];
+
+	/* Observale that will display loaded packages from NPM */
+	readonly autocompleteOptions$!: Observable<string[]>;
+
+	/* Form control to allow user search packages  */
+	readonly addPackage: FormControl<string> = new FormControl('', {
+		validators: this.errorHandlerService.noDuplicatesValidator(this.alreadyLoadedPackageName),
+		nonNullable: true,
+	});
+
+	/* errors that may occur when searching for a package name */
+	readonly packageErrorsHandler = this.errorHandlerService.getInputErrorsHandler('package name');
+
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	onChange = (value: string[]) => {
+		/* empty */
+	};
+	onTouched = () => {
+		/* empty */
+	};
+	constructor(
+		private readonly npmRegistryService: NpmRegistryService,
+		private readonly errorHandlerService: ErrorHandlerService
+	) {
+		// npm library suggestions on user input
+		const suggestions$ = this.searchLibraryOnInputChange();
+
+		this.autocompleteOptions$ = combineLatest([
+			// All suggestions
+			suggestions$.pipe(startWith([])),
+			// Partial npm package name to filter options
+			this.addPackage.valueChanges.pipe(startWith('')),
+		]).pipe(
+			map(([suggestions, query]) =>
+				this.getAutocompleteOptions(
+					[...new Set([...(this.autocompletePackageNames ?? []), ...suggestions])],
+					this.alreadyLoadedPackageName,
+					query
+				)
+			)
+		);
+	}
+
+	writeValue(value?: string[]): void {
+		this.alreadyLoadedPackageName = [...(value ?? [])];
+	}
+	registerOnChange(fn: (value: string[]) => void): void {
+		this.onChange = fn;
+	}
+	registerOnTouched(fn: (value: void) => void): void {
+		this.onTouched = fn;
+	}
+
+	validate(c: AbstractControl): ValidationErrors | null {
+		return c.validator;
```

Comment:
here we would pass a custom validator: 
```typescript
return someValidator(c);
```

Reply:
what validation should I pass there, this one
`this.errorHandlerService.noDuplicatesValidator(this.alreadyLoadedPackageName)` ?

Reply:
fixed in [10c7030](https://github.com/bitovi/opensource-analytics-dashboard/pull/9/commits/10c7030a07fc521fc8197d5356fa58516c23eb77)